viewer/arvviewer.c

```diff
@@ -241,8 +241,30 @@ arv_viewer_value_from_log (double value, double min, double max)
 	return pow (10.0, (value * (log10 (max) - log10 (min)) + log10 (min)));
 }
 
+struct free_arv_buffer_cb_param
+{
+	GWeakRef stream;
+	GWeakRef arv_buffer;
+};
+
+static void
+free_arv_buffer_cb (void* param)
+{
+	struct free_arv_buffer_cb_param* cb_param = (struct free_arv_buffer_cb_param*)param;
+
+	ArvStream* stream = g_weak_ref_get (&cb_param->stream);
+	ArvBuffer* arv_buffer = g_weak_ref_get (&cb_param->arv_buffer);
```

Comment:
arv_buffer should always be valid in this callback, so using g_weak_ref on it is not needed. You should be careful to unref it in the case arv_stream_push_buffer is not called.